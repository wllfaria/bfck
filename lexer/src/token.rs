/// brainfuck has only 8 commands, those are all described below, you can refer
/// to each of them in more detail following the link below
///
/// https://en.wikipedia.org/wiki/Brainfuck
#[derive(Debug, Clone, Copy)]
pub enum Token {
    /// move (decrement) the data pointer by one, essentially pointing to the
    /// next cell to the left
    MoveLeft(usize),
    /// move (increment) the data pointer by one, essentially pointing to the
    /// next cell to the right
    MoveRight(usize),
    /// increment the byte at the data pointer by one, as the specification
    /// mentions, this wraps when `u8::MAX` is reached, going back to 0
    Increment(usize),
    /// decrement the byte at the data pointer by one, as the specification
    /// mentions, this wraps when `0` is reached, going back to `u8::MAX`
    Decrement(usize),
    /// outputs the byte at the data pointer
    Write(usize),
    /// reads one byte from stdin
    Read(usize),
    /// if the byte at the data pointer is zero, then instead of moving the
    /// instruction pointer to the next instruction, move it to the instruction
    /// right after the matching closing ´]´
    JumpIfZero(usize),
    /// if the byte at the data pointer is non zero, then instead of moving the
    /// instruction pointer to the next instruction, move it back to the
    /// instruction right after the matching opening ´[´
    JumpUnlessZero(usize),
}

/// this custom partial equality check exists as for this implementation, we
/// don't care about the inner value, which is the count of times this token
/// was repeated. We only care if both tokens are the same, so we can match
/// any new token which will always have count 1, to an existing token and
/// as long as they are the same type, we can increment the ineer count.
///
/// this is how we reduce the tokens to a smaller amount by grouping by
/// repetition
impl PartialEq for Token {
    fn eq(&self, other: &Self) -> bool {
        matches!(
            (self, other),
            (Token::MoveLeft(_), Token::MoveLeft(_))
                | (Token::MoveRight(_), Token::MoveRight(_))
                | (Token::Increment(_), Token::Increment(_))
                | (Token::Decrement(_), Token::Decrement(_))
                | (Token::Write(_), Token::Write(_))
                | (Token::Read(_), Token::Read(_))
                | (Token::JumpIfZero(_), Token::JumpIfZero(_))
                | (Token::JumpUnlessZero(_), Token::JumpUnlessZero(_))
        )
    }
}

impl Token {
    /// helper function to get the inner count as a mutable reference, this is
    /// used to increment the count whenever we found a repeated tokens in a
    /// sequence
    pub fn inner(&mut self) -> &mut usize {
        match self {
            Token::MoveLeft(count) => count,
            Token::MoveRight(count) => count,
            Token::Increment(count) => count,
            Token::Decrement(count) => count,
            Token::Write(count) => count,
            Token::Read(count) => count,
            Token::JumpIfZero(count) => count,
            Token::JumpUnlessZero(count) => count,
        }
    }
}

/// when tokenizing the input, we convert the string slice into a byte array,
/// which is easier to handle while keeping the tokenizer 0 copy. this is a
/// convenience impl that makes it easier to find out wether we should skip
/// the token when its not any of the valid tokens, or collect it into a
/// existing, or new group of tokens
impl TryFrom<u8> for Token {
    type Error = &'static str;

    fn try_from(value: u8) -> Result<Token, Self::Error> {
        match value {
            b'<' => Ok(Token::MoveLeft(1)),
            b'>' => Ok(Token::MoveRight(1)),
            b'+' => Ok(Token::Increment(1)),
            b'-' => Ok(Token::Decrement(1)),
            b'.' => Ok(Token::Write(1)),
            b',' => Ok(Token::Read(1)),
            b'[' => Ok(Token::JumpIfZero(1)),
            b']' => Ok(Token::JumpUnlessZero(1)),
            _ => Err("not a valid token"),
        }
    }
}

impl std::fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Token::MoveLeft(count) => write!(f, "{}<", count),
            Token::MoveRight(count) => write!(f, "{}>", count),
            Token::Increment(count) => write!(f, "{}-", count),
            Token::Decrement(count) => write!(f, "{}+", count),
            Token::Write(count) => write!(f, "{}.", count),
            Token::Read(count) => write!(f, "{},", count),
            Token::JumpIfZero(count) => write!(f, "{}[", count),
            Token::JumpUnlessZero(count) => write!(f, "{}]", count),
        }
    }
}
