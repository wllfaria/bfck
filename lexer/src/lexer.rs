use crate::token::Token;

/// Lexer for the brainfuck source code, exposes only one method, namely `tokenize`.
/// The responsibility of the lexer is to receive a code string and tokenize however
/// many tokens are available into semantic tokens.
///
/// This lexer also reduces the tokens to a intermediate representation, where
/// instead of having a `Token` for possibly every byte in the source code, we reduce
/// it to an intermediate representation that collapses repeating tokens into a single
/// one.
///
/// example:
/// given the brainfuck source code `>>>-----..`; the lexer would produce a vector like
/// the following
/// ```rust
/// use lexer::Lexer;
/// use lexer::token::Token;
///
/// let source = ">>>-----..";
/// let expected = vec![
///     Token::MoveRight(3),
///     Token::Decrement(5),
///     Token::Write(2),
/// ];
///
/// let tokens = Lexer::tokenize(source);
/// assert_eq!(tokens, expected);
/// ```
pub struct Lexer {}

impl Lexer {
    /// given a input slice of source code, Lexer will try to tokenize it ignoring all
    /// whitespaces and non-tokens and always give back a vector of tokens, which
    /// could be empty if no valid tokens exists within the input string
    ///
    /// no validations are performed here purposefully, its not the responsibility of
    /// the lexer to validate if the source code is valid.
    pub fn tokenize(input: &str) -> Vec<Token> {
        let input = input.as_bytes();
        let mut tokens = vec![];
        let mut index = 0;

        while index < input.len() {
            let token = Token::try_from(input[index]);

            // we couldn't convert the byte to a valid token, so we just consider it
            // a comment and skip over it
            if token.is_err() {
                index += 1;
                continue;
            }

            // its obviously safe to unwrap here since we check above, but lets make
            // it clear this is intended!
            assert!(token.is_ok(), "token was not valid yet we didn't skip it");
            let mut token = token.unwrap();

            // since we found a valid token, we will keep collecting tokens until we
            // find a different one
            index += 1;
            while index < input.len() {
                let next = Token::try_from(input[index]);

                // again, if the byte is a non-valid brainfuck token, we just skip
                // over it
                if next.is_err() {
                    index += 1;
                    continue;
                }

                // if it is a valid token, but its a different one from the one we
                // initially collected, we break out of this loop without
                // incrementing the index, so the next outer loop iteration will
                // resume from there
                //
                // its obviously safe to unwrap here since we check above, but lets make
                // it clear this is intended!
                assert!(next.is_ok(), "token was not valid yet we didn't skip it");
                if next.unwrap().ne(&token) {
                    break;
                }

                // if the tokens are the same, we increase the count of the token
                // through the inner() helper and increase the index to we do the
                // same with the next
                *token.inner() += 1;
                index += 1;
            }

            tokens.push(token);
        }

        tokens
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn check(input: &'static str) -> Vec<Token> {
        Lexer::tokenize(input)
    }

    #[test]
    fn empty_input() {
        insta::assert_debug_snapshot!(check("     "));
    }

    #[test]
    fn single_spaced_token() {
        insta::assert_debug_snapshot!(check("<<       <<"));
    }

    #[test]
    fn changing_tokens() {
        insta::assert_debug_snapshot!(check("<<  >>"));
    }

    #[test]
    fn ignoring_non_tokens() {
        insta::assert_debug_snapshot!(check(" _*  <<random_!?anything!!>!!>"));
    }

    #[test]
    fn pretty_hello_world() {
        insta::assert_debug_snapshot!(check(include_str!("../../samples/hello_world_pretty.bf")));
    }
}
