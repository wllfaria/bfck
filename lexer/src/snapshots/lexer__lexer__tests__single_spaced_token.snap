---
source: lexer/src/lexer.rs
expression: "check(\"<<       <<\")"
---
[
    MoveLeft(
        4,
    ),
]
